tool/tsh/azure.go

```diff
@@ -0,0 +1,423 @@
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+const (
+	azureCLIBinaryName = "az"
+)
+
+func onAzure(cf *CLIConf) error {
+	app, err := pickActiveAzureApp(cf)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = app.StartLocalProxies()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	defer func() {
+		if err := app.Close(); err != nil {
+			log.WithError(err).Error("Failed to close Azure app.")
+		}
+	}()
+
+	args := cf.AzureCommandArgs
+
+	cmd := exec.Command(azureCLIBinaryName, args...)
+	return app.RunCommand(cmd)
+}
+
+// azureApp is an Azure app that can start local proxies to serve Azure APIs.
+type azureApp struct {
+	cf      *CLIConf
+	profile *client.ProfileStatus
+	app     tlsca.RouteToApp
+
+	localALPNProxy    *alpnproxy.LocalProxy
+	localForwardProxy *alpnproxy.ForwardProxy
+}
+
+// newAzureApp creates a new Azure app.
+func newAzureApp(cf *CLIConf, profile *client.ProfileStatus, app tlsca.RouteToApp) (*azureApp, error) {
+	return &azureApp{
+		cf:      cf,
+		profile: profile,
+		app:     app,
+	}, nil
+}
+
+// StartLocalProxies sets up local proxies for serving Azure clients.
+//
+// There are two ways clients can connect to the local proxies.
+//
+// 1. client can send Azure requests to our local forward proxy by configuring
+// HTTPS_PROXY (or equivalent). The API flow looks like this:
+// clients -> local forward proxy -> local ALPN proxy -> remote server
+//
+// 2. client can send Azure requests to our local ALPN proxy directly by
+// configuring Azure endpoint URLs. The API flow looks like this.
+// clients -> local ALPN proxy -> remote server
+//
+// The first method is always preferred as the original hostname is preserved
+// through forward proxy.
+func (a *azureApp) StartLocalProxies() error {
+	// HTTPS proxy mode
+	if err := a.startLocalALPNProxy(""); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := a.startLocalForwardProxy(a.cf.LocalProxyPort); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// Close makes all necessary close calls.
+func (a *azureApp) Close() error {
+	var errs []error
+	if a.localALPNProxy != nil {
+		errs = append(errs, a.localALPNProxy.Close())
+	}
+	if a.localForwardProxy != nil {
+		errs = append(errs, a.localForwardProxy.Close())
+	}
+	return trace.NewAggregate(errs...)
+}
+
+// GetEnvVars returns required environment variables to configure the
+// clients.
+func (a *azureApp) GetEnvVars() (map[string]string, error) {
+	homeDir, err := os.UserHomeDir()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	envVars := map[string]string{
+		// set custom Azure home path; this helps with the scenario in which user runs
+		// 1. `tsh az login` in one console
+		// 2. `az ...` in another console
+		// without custom config dir the second invocation will hang, attempting to connect to (inaccessible without configuration) MSI.
+		"AZURE_CONFIG_DIR": path.Join(homeDir, ".azure-teleport-"+a.app.Name),
```

Comment:
I don't think we should remove these files even if they are stale. Among other things az CLI stores user-provided configuration (see: https://learn.microsoft.com/en-us/cli/azure/azure-cli-configuration). If we were to wipe out that folder on logout it would naturally be lost.

For the same reason I don't think we should use `~/.tsh` as a parent directory.